### PR TITLE
[Bugfix] Fix support for dimension like integers and ScalarType

### LINKED ISSUE #9234: [issue tracker] make quantization compatible with dynamo dynamic shape
### Anything you want to discuss about vllm.

here is a simple demo code:

```python
import torch
from torch.utils.cpp_extension import load_inline

custom_library = torch.library.Library("custom", "DEF")
custom_library.define("add_cpp(Tensor x, int y) -> Tensor")

cpp_source = """
#include <torch/extension.h>

torch::Tensor custom_add(torch::Tensor x, int64_t y) {
    return x + y;
}

TORCH_LIBRARY_IMPL(custom, CPU, m) {
    m.impl("add_cpp", custom_add);
}
"""

custom_op = load_inline(
    name="custom_op",
    cpp_sources=cpp_source,
    extra_cflags=[],
    functions=["custom_add"]
)

@torch.library.register_fake("custom::add_cpp")
def _(x: torch.Tensor, y: int) -> torch.Tensor:
    return torch.empty((y,), dtype=torch.float32)

import torch

@torch.library.custom_op("custom::add_py", mutates_args=[])
def add_py(x: torch.Tensor, y: int) -> torch.Tensor:
    return x + y

@add_py.register_fake
def _(x: torch.Tensor, y: int) -> torch.Tensor:
    return torch.empty((y,), dtype=torch.float32)

@torch.compile(backend="eager", fullgraph=True)
def f(x):
    # return torch.ops.custom.add_py(x, x.shape[0]) # passes
    return torch.ops.custom.add_cpp(x, x.shape[0]) # errors with `Not all values of RelaxedUnspecConstraint(L['x'].size()[0]) are valid because L['x'].size()[0] was inferred to be a constant (2).`

x = torch.ones(2, 4)
torch._dynamo.mark_dynamic(x, 0)
print(f(x)[0])
```

when we register the custom op from c++ side, dynamic shape will be directly specialized to an integer, and fail.
when we register the custom op from Python side, dynamic shape works as expected.

we should change the way we register quantization as custom ops, from c++ side to python side. 

there's also one complicated object https://github.com/vllm-project/vllm/blob/f3a507f1d31e13a99c4fc8ac02738a73c3e3136f/vllm/scalar_type.py#L15 that appears in the custom op parameter : 

https://github.com/vllm-project/vllm/blob/f3a507f1d31e13a99c4fc8ac02738a73c3e3136f/vllm/_custom_ops.py#L315-L321

we can use strings to represent the type, and look up the actual object to pass into the c++ function.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by bnellnm**
Some of the custom ops take integers which are `Tensor` dimensions.  These can sometimes be `SymInt`s if those `Tensor` dimensions are marked as dynamic.  The appropriate `int` arguments for these class of custom ops have been changed to `SymInt` to support dynamic dimensions.

The inductor is not currently able to support passing custom C++ classes to custom ops.  This PR fully implements `ScalarType` in python.  `ScalarType`s are now passed by `id` to C++ where they are reconstructed into C++ `ScalarType`s.  This has the side effect of removing the `_core_C` extension.

fixes https://github.com/vllm-project/vllm/issues/9234

cc @SageMoore , @ProExpertProg , @LucasWilkinson , @youkaichao 

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/scalar_type.py [hunk 1]
```diff
@@ -1,4 +1,284 @@
...
+            assert (not self.is_signed() or
+                    self.size_bits <= 64), "Cannot represent min as a int64_t"
+
+            if self.is_signed():
+                return -(1 << (self.size_bits - 1))
+            else:
+                return 0
+
+    @property
+    def id(self) -> int:
```
Comment (LucasWilkinson) [inline vllm/scalar_type.py:132]: can you add a comment that this should be kept up to date with `from_id` and `id` in `scalar_type.hpp`

**Comment by LucasWilkinson** [comment]
Thanks for doing this! Sorry scalar type caused this issue, stinks that inductor doesnt support custom classes. But more red then green in this PR so an overall win I guess :+1:. I am a bit worried about python based `id` function calls on the hot path as the id function can be a bit slow in python (im seeing about 14us)
Screenshot 2024-10-11 at 11 57 25 PM
although we do appear to saving an expensive `_must_dispatch_in_python` which appears to more than offset that, 
Screenshot 2024-10-12 at 12 04 30 AM
but might still be good to lift the `id` call out of `_custom_ops.py` so users can optimize that if they please

**Comment by bnellnm** [comment]
> Thanks for doing this! Sorry scalar type caused this issue, stinks that inductor doesnt support custom classes. But more red then green in this PR so an overall win I guess 👍. I am a bit worried about python based `id` function calls on the hot path as the id function can be a bit slow in python (im seeing about 14us) Screenshot 2024-10-11 at 11 57 25 PM although we do appear to saving an expensive `_must_dispatch_in_python` which appears to more than offset that, Screenshot 2024-10-12 at 12 04 30 AM but might still be good to lift the `id` call out of `_custom_ops.py` so users can optimize that if they please

I've memoized `id` and made sure it is populated when the types are constructed.

**Comment by LucasWilkinson** [comment]
I was using:
```
python examples/offline_profile.py --model neuralmagic/Meta-Llama-3.1-8B-Instruct-quantized.w4a16 --batch-size 4 --prompt-len 512 --json trace_test --save-chrome-traces-folder chrome_traces --max-num-batched-tokens 8196
```
off this branch: https://github.com/vllm-project/vllm/pull/8337

Then opening `profill.json` in `chrome_traces` using https://ui.perfetto.dev/, you have to merge the branches though so I just went ahead and re-ran it here:

image

looks good! :+1: not seeing any overhead, thanks!

File: vllm/scalar_type.py [hunk 2]
```diff
@@ -1,4 +1,298 @@
...
+    """
+    bias used to encode the values in this scalar type
+    (value = stored_value - bias, default 0) for example if we store the
+    type as an unsigned integer with a bias of 128 then the value 0 will be
+    stored as 128 and -1 will be stored as 127 and 1 will be stored as 129.
+    """
+
+    _finite_values_only: bool = False
+    """
+    Private: if NANs are supported, used `has_infs()` instead.
```
Comment (LucasWilkinson) [inline vllm/scalar_type.py:58]: this should be `if infs are supported` (as opposed to `if NANs are supported`), sorry the original comment was buggy 😞

File: vllm/scalar_type.py [hunk 3]
```diff
@@ -1,4 +1,298 @@
...
+        # complication we are just assuming the standard exponent bias until
+        # there is a need to support non-standard biases
+        exponent_bias = (1 << (self.exponent - 1)) - 1
+        exponent_bias_double = (1 << 10) - 1  # double e = 11
+
+        max_exponent_double = (max_exponent - exponent_bias +
+                               exponent_bias_double)
+
+        # shift the mantissa into the position for a double and
+        # the exponent
```
Comment (LucasWilkinson) [inline vllm/scalar_type.py:98]: maybe update this to:
```
shift the mantissa and the exponent into the position for a double, then or them together
```

reads a bit funny right now (may have been my bad if the original comment was bad)

**Review by ProExpertProg — APPROVED** [review]

File: csrc/torch_bindings.cpp [hunk 4]
```diff
@@ -201,8 +196,8 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, ops) {
   ops.def(
       "gptq_marlin_gemm(Tensor a, Tensor b_q_weight, Tensor b_scales, "
       "Tensor b_zeros, Tensor g_idx, Tensor perm, Tensor workspace, "
-      "__torch__.torch.classes._core_C.ScalarType b_q_type, "
-      "int size_m, int size_n, int size_k, bool is_k_full, "
+      "int b_q_type, "
+      "SymInt size_m, SymInt size_n, SymInt size_k, bool is_k_full, "
```
[the remark this answers is not in the document]
Comment (bnellnm) [reply csrc/torch_bindings.cpp:200]: Essentially, yeah.

**Review by SageMoore — APPROVED** [review]

File: tests/compile/utils.py [hunk 5]
```diff
@@ -69,11 +69,11 @@ def check_full_graph_support(model,
...
     os.environ["VLLM_TEST_DYNAMO_FULLGRAPH_CAPTURE"] = "1"
 
-    # Inductor doesn't support fp8/gptq_marlin_24 yet.
+    # Inductor doesn't support fp8 and the base meta llama uses too
+    # much memory.
     quantization = model_kwargs.get("quantization")
-    if (quantization == "fp8" or quantization == "gptq_marlin"
-            or quantization == "gptq_marlin_24"
-        ) and optimization_level >= CompilationLevel.INDUCTOR:
+    if ((quantization == "fp8" or model == "meta-llama/Meta-Llama-3-8B")
```
[the remark this answers is not in the document]
Comment (bnellnm) [reply tests/compile/utils.py:75]: I'd rather the check be specific to the problem model(s) rather than making it more generic, e.g. once we upgrade to 2.5 we can remove the `fp8` check.  It's possible this llama model will also run without oom'ing.

File: vllm/_custom_ops.py [hunk 6]
```diff
@@ -475,19 +478,6 @@ def causal_conv1d_update_fake(
...
-                                A: torch.Tensor, B: torch.Tensor,
-                                C: torch.Tensor, D_: Optional[torch.Tensor],
-                                z_: Optional[torch.Tensor],
-                                delta_bias_: Optional[torch.Tensor],
-                                delta_softplus: bool,
-                                cu_seq_len: Optional[torch.Tensor],
-                                cache_indices: Optional[torch.Tensor],
-                                has_initial_state: Optional[torch.Tensor],
-                                ssm_states: Optional[torch.Tensor]) -> None:
-        return None
```
Comment (tlrmchlsmth) [inline vllm/_custom_ops.py:489]: how come this goes away?

**Review by tlrmchlsmth — APPROVED** [review]
RIP ScalarType

File: csrc/moe/torch_bindings.cpp [hunk 7]
```diff
@@ -13,8 +13,9 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, m) {
       "marlin_gemm_moe(Tensor! a, Tensor! b_q_weights, Tensor! sorted_ids, "
       "Tensor! topk_weights, Tensor! topk_ids, Tensor! b_scales, Tensor! "
       "b_zeros, Tensor! g_idx, Tensor! perm, Tensor! workspace, "
-      "__torch__.torch.classes._core_C.ScalarType b_q_type, int size_m, "
-      "int size_n, int size_k, bool is_k_full, int num_experts, int topk, "
+      "int b_q_type, SymInt size_m, "
+      "SymInt size_n, SymInt size_k, bool is_k_full, int num_experts, int "
```
Comment (tlrmchlsmth) [inline csrc/moe/torch_bindings.cpp:17]: Should `size_n` and `size_k` be regular ints since they will be known at torch.compile time?

Comment (tlrmchlsmth) [reply csrc/moe/torch_bindings.cpp:17]: Ditto for all `N`s and `K`s

**Review by youkaichao — APPROVED** [review]
thanks for the great efforts! please address comments from @tlrmchlsmth ?

**Comment by tlrmchlsmth** [comment]
> thanks for the great efforts! please address comments from @tlrmchlsmth ?

From an offline discussion with @bnellnm, it sounds like there's no downside to leaving N and K as SymInts -- good to go from my end

File: vllm/_custom_ops.py (hunk 6 above)
Comment (bnellnm) [reply vllm/_custom_ops.py:489]: fake functions aren't required for nullary ops.
